### DISCUSSION TITLE
Example Gallery: Strategy and Reasoning

### DISCUSSION THREAD

**Discussion by michaelweinold** [brightway-lca/brightway-documentation#116] [Strategy and Planning]
Add a gallery like in e.g. [`pvlib`](https://pvlib-python.readthedocs.io/en/stable/gallery/index.html). Potential packages: [`sphinx-gallery`](https://github.com/sphinx-gallery/sphinx-gallery), [`nbsphinx`](https://github.com/spatialaudio/nbsphinx/), [`myst-nb`](https://github.com/executablebooks/myst-nb).
See also [this comparison](https://www.githubcompare.com/executablebooks/myst-nb+sphinx-gallery/sphinx-gallery+spatialaudio/nbsphinx).
Excellent example: [ArviZ documentation example gallery](https://python.arviz.org/en/stable/examples/index.html)

**Comment by michaelweinold** [comment]
`sphinx-gallery`/`nbsphinx` rely on each other to generate a gallery. However, they copy notebooks to an autogenerated directory that further complicates the build process and makes it nearly impossible to include images, etc.

Compary my issue here:

Embedding images in `rst` blocks currently broken, compare: 
 - [x] https://github.com/sphinx-gallery/sphinx-gallery/issues/1063

We will __not__ be using these packages.

**Comment by michaelweinold** [comment]
`myst-nb` package:

This package is by far the most powerful when rendering Jupyter Notebooks with embedded markdown. It is also part of the `executablebooks` project. As of now, it does __not__ include a directive for generating a thumbnail gallery. However, multiple related issues are already open:

- [ ] https://github.com/executablebooks/MyST-NB/issues/396
- [ ] https://github.com/PlasmaPy/PlasmaPy/pull/1474
- [ ] https://github.com/spatialaudio/nbsphinx/issues/655

We will wait until `myst-nb` supports a thumbnail gallery before adding it to the documentation.

**Comment by michaelweinold** [comment]
For now, the `grid` directive of the `myst-parser` is by far good enough for the example gallery.

This means that for now, the notebooks are __not__ executed upon build and the gallery is __not__ built automatically. However, it also means that building the gallery can be accomplished in `myst`-markdown alone. Also, no additional packages are required and the build does not become a complete mess.

**Comment by michaelweinold** [comment]
Tradeoffs:

| package(s) | + | - | comment |
| -------- | ------- | -------- | ----------|
| [`sphinx-gallery`](https://sphinx-gallery.github.io/stable/index.html) | download example as `ipynb` <br>  can execute notebooks | example source must be `.py` <br> no Markdown support | generates an autobuild folder |
| [`nbsphinx`+`sphinx-gallery`](https://nbsphinx.readthedocs.io/en/0.8.12/subdir/gallery.html) | markdown support (`.mystnb`) <br> can execute notebooks | no simple download option (possible through ["prolog/epilog"](https://github.com/spatialaudio/nbsphinx/issues/5) | 
| `myst-nb` | Markdown support | requires manual creation of gallery <br> c̶a̶n̶n̶o̶t̶ e̶x̶e̶c̶u̶t̶e̶ n̶o̶t̶e̶b̶o̶o̶k̶s̶ |

**Comment by michaelweinold** [comment]
# Decision

1. use `myst-nb` for now, with a manual grid of examples
2. wait until https://github.com/executablebooks/MyST-NB/issues/396 has been implemented, then use this new functionality to auto-generate a gallery of examples
3. wait until https://github.com/executablebooks/MyST-NB/issues/148 is implemented, then use this new functionality to add a download button to the notebooks

**Comment by michaelweinold** [comment]
...the PyMC documentation has their examples in a separate repository: [pymc-examples](https://github.com/pymc-devs/pymc-examples)

- [ ] are they using `myst-nb` or `Sphinx-Gallery`?
- [ ] how are they integrating the examples into the main documentation

[conf.py](https://github.com/pymc-devs/pymc/blob/1ed447573261d0b0e7306a95962f2d2d9ae9f171/docs/source/conf.py#L198)
```
"nb": ("https://www.pymc.io/projects/examples/en/latest/", None),
```

...currently, they are using their own script: [pymc-examples/sphinxext/thumbnail_extractor.py](https://github.com/pymc-devs/pymc-examples/blob/ea709bbca6c53618be66fcfe5434164081cd68eb/sphinxext/thumbnail_extractor.py#L35)

**Comment by michaelweinold** [comment]
- [ ] https://github.com/executablebooks/MyST-NB/issues/396#issuecomment-1514945851

**Comment by michaelweinold** [comment]
https://myst-nb.readthedocs.io/en/latest/authoring/text-notebooks.html

**Comment by cmutel** [comment]
@michaelweinold Could you add a brief note on the advantage of these MyST notebooks versus standard notebooks (sorry if this is obvious to you but it wasn't to me).

I also see an easy conversion to Jupyter Notebooks, but can we also convert the other way. We don't really want to rewrite working example notebooks unless absolutely necessary.

**Comment by michaelweinold** [comment]
@cmutel, the primary advantage is better diffing. While GitHub will likely release their current [Jupyter Notebook Diff feature preview](https://github.blog/changelog/2023-03-01-feature-preview-rich-jupyter-notebook-diffs/) later this year, this would still leave local editors at a disadvantage. Working with `.ipynb` files can be a pain.

In the bigger picture, I doubt if we can transfer the present notebooks to either the JupterBook site or the JupyterLite hub site without 'renovating' them.

This issue here is helping me [collect information on the two options we have](https://github.com/brightway-lca/brightway-documentation/issues/45#issuecomment-1399416846) to use Jupyter Notebooks with the Sphinx documentation.

**Comment by michaelweinold** [comment]
My discussion in the Executablebooks repo: [`sphinx-design` multiple Grids with same Height](https://github.com/orgs/executablebooks/discussions/1060)

Potentially relevant: 

https://sphinx-design.readthedocs.io/en/pydata-theme/css_classes.html#avatars
https://sphinx-design.readthedocs.io/en/pydata-theme/css_classes.html#sizing
